R/bundle.R

```diff
@@ -1119,3 +1147,59 @@ performPackratSnapshot <- function(bundleDir, verbose = FALSE) {
   # TRUE just to indicate success
   TRUE
 }
+
+# Run "quarto inspect" on the target and returns its output as a parsed object.
+quartoInspect <- function(appDir = NULL, appPrimaryDoc = NULL, quarto = NULL) {
+  if (!is.null(quarto)) {
+    inspect <- NULL
+    # If "quarto inspect appDir" fails, we will try "quarto inspect
+    # appPrimaryDoc", so that we can support single files as well as projects.
+    primaryDocPath <- file.path(appDir, appPrimaryDoc) # prior art: appHasParameters()
+    for (path in c(appDir, primaryDocPath)) {
+      args <- c("inspect", path.expand(path))
+      inspect <- suppressWarnings(system2(quarto, args, stdout = TRUE, stderr = FALSE))
+      if (is.null(attr(inspect, "status"))) {
+        return(jsonlite::fromJSON(inspect))
+      }
+    }
+  }
+  return(NULL)
+}
+
+# Extract the Quarto version and engines, either from a parsed "quarto inspect"
+# result or from metadata provided by the quarto-r package.
+getQuartoManifestDetails <- function(inspect = list(), metadata = list()) {
+  if (length(inspect) != 0) {
+    return(list(
+      version = inspect[["quarto"]][["version"]],
+      engines = I(inspect[["engines"]])
+    ))
+  }
+  if (!is.null(metadata[["quarto_version"]])) {
+    return(list(
+      "version" = metadata[["quarto_version"]],
+      "engines" = metadata[["quarto_engines"]]
+    ))
+  }
+  return(NULL)
+}
+
+# Attempt to gather Quarto version and engines, first from quarto inspect if a
+# quarto executable is provided, and then from metadata.
+inferQuartoInfo <- function(appDir, appPrimaryDoc, quarto, metadata = NULL) {
+  quartoInfo <- NULL
+  if (!is.null(quarto)) {
+    # Prefer user-provided Quarto path over metadata
+    inspect <- quartoInspect(
+      appDir = appDir,
+      appPrimaryDoc = appPrimaryDoc,
+      quarto = quarto
+    )
+    quartoInfo <- getQuartoManifestDetails(inspect = inspect)
+  }
+  if (is.null(quartoInfo) && !is.null(metadata)) {
+    # If we don't yet have Quarto details, attempt to extract from metadata
+    quartoInfo <- getQuartoManifestDetails(metadata = metadata)
```

Comment:
If we already have Quarto information in the metadata, it means that someone has already called `quarto inspect` and we do not need to, no?

Reply:
What I was trying to say: When `metadata` has `quarto_version` and `quarto_engines`, we should use that information and avoid calling `quarto inspect`.